### ISSUE TITLE
strictTemplates and strictNullChecks doesn't work well with optional chaining in templates

### ISSUE THREAD

**Issue by ajafff** [angular/angular#37619]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/compiler?

### Is this a regression?

?

### Description

* enable Ivy
* enable `strictTemplates`
* enable `strictNullChecks`
* use an expression like `foo?.bar.baz` in a template
  * `foo` is nullable, `bar` is not
  * template compiler shows error on `baz`: `Object is possibly 'undefined'.`
* note that the whole chain ends if `foo` is nullish (long short circuiting), therefore accessing `baz` is always safe
* disabling `strictNullChecks` makes the compile error go away and everything functions as expected without runtime error (i.e. Angular's optional chaining also does long short circuiting)

## 🔬 Minimal Reproduction
https://stackblitz.com/edit/angular-ivy-pu8dk5

## 🔥 Exception or Error
<pre><code>
`Object is possibly 'undefined'.`
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
@angular/animations 9.1.11
@angular/common 9.1.11
@angular/compiler 9.1.11
@angular/core 9.1.11
@angular/forms 9.1.11
@angular/platform-browser 9.1.11
@angular/platform-browser-dynamic 9.1.11
@angular/router 9.1.11

</code></pre>

**Anything else relevant?**

**Comment by ajafff** [comment]
seems to be fixed with 10.0.0-rc.6

**Comment by ajafff** [comment]
I take that back. It seems that stackblitz doesn't report any compile errors with Angular 10.0.0-rc.6, so this issue seems to still exist in that version

**Comment by JoostK** [comment]
Yes, this is an issue. We generate:

```ts
    "" + ((((null as any) ? (((ctx).prop /*24,28*/) /*24,28*/)!.nested : undefined) /*24,36*/).field /*37,42*/) /*24,42*/;
```

which is obviously broken.

A workaround could be to use `!` for subsequent property reads, or set `strictSafeNavigationTypes` to false (that will result in less accurate checking, as the expression is interpreted as `any`)

**Comment by ajafff** [comment]
adding `!` ends the optional chain, preventing long short circuiting, causing a runtime error, because `field` is always accessed even if `prop` is nullish. See my related comment at https://github.com/angular/angular/issues/34385#issuecomment-645321811

**Comment by JoostK** [comment]
@ajafff Oh, interesting! Thanks for bringing that up.

**Comment by JoostK** [comment]
@ajafff I think it's worth capturing that as an issue by itself, regardless of how #34385 is dealt with.

**Comment by PowerKiKi** [comment]
Still seeing this on Angular 10.2 on a brand new strict project. Adding `"strictSafeNavigationTypes": false` in `tsconfig.json` is the only reasonable way to adopt other strict rules for now. But that is not ideal since we lose one kind of rule.

Here is as reproduction case on Angular 10.2. In essence it is very similar to @ajafff's repro case:

https://stackblitz.com/edit/strictsafenavigationtypes-is-broken?file=src%2Fapp%2Fapp.component.html

**Comment by manueledones** [comment]
still an issue on angular 11, I'm bypassing it without using optional chaining, like `*ngIf="foo && foo.bar.baz"`

**Comment by mvnrhd** [comment]
still in 10.2.4, hoping for a soon fix 🙏

**Comment by klofi** [comment]
Still broken in new project created with `@angular/cli` v11.1.4. Getting around by using `"strictNullChecks": false` until fixed.

**Comment by amakhrov** [comment]
It looks like the issue only concerns an incorrectly generated type check block - the actual emitted JS is fine. So perhaps fixing the type checker would be a non-breaking change for a next major (or even a minor) version - comparing to a larger exercise in https://github.com/angular/angular/issues/34385 ?

**Comment by sod** [comment]
Reproduction in angular 12.1 https://github.com/sod/ng-strict-issue1

Right now as soon as you use the operator in the template, angular forces you to use it everywhere in that statement, even though, nested structures may be guaranteed to not be undefined by the types.

**Comment by ties-s** [comment]
This has been over a year. This is a very annoying bug. Any updates on the progress? This shouldn't take more than 2 major versions.

**Comment by amakhrov** [comment]
@JoostK 
> Yes, this is an issue. We generate:
> ```ts
>     "" + ((((null as any) ? (((ctx).prop /*24,28*/) /*24,28*/)!.nested : undefined) /*24,36*/).field /*37,42*/) /*24,42*/;
> ```
> 

Is there a reason we cannot generate TCB with native `?.` syntax inside?
Basically, instead of
```
const expr = ts.createPropertyAccess(ts.createNonNullExpression(receiver), ast.name);
// result: ctx.prop!.nested

node = ts.createParen(ts.createConditional(NULL_AS_ANY, expr, UNDEFINED));
// result: (null as any ? ctx.prop!.nested : undefined)
```
do
```
node = ts.createPropertyAccessChain(receiver, ts.createToken(SyntaxKind.QuestionDotToken), ast.name);
// result: ctx.prop?.nested
```

**Comment by JoostK** [comment]
> @JoostK 
> 
> > Yes, this is an issue. We generate:
> 
> > ```ts
> 
> >     "" + ((((null as any) ? (((ctx).prop /*24,28*/) /*24,28*/)!.nested : undefined) /*24,36*/).field /*37,42*/) /*24,42*/;
> 
> > ```
> 
> > 
> 
> 
> 
> Is there a reason we cannot generate TCB with native `?.` syntax inside?
> 
> Basically, instead of
> 
> ```
> 
> const expr = ts.createPropertyAccess(ts.createNonNullExpression(receiver), ast.name);
> 
> // result: ctx.prop!.nested
> 
> 
> 
> node = ts.createParen(ts.createConditional(NULL_AS_ANY, expr, UNDEFINED));
> 
> // result: (null as any ? ctx.prop!.nested : undefined)
> 
> ```
> 
> do
> 
> ```
> 
> node = ts.createPropertyAccessChain(receiver, ts.createToken(SyntaxKind.QuestionDotToken), ast.name);
> 
> // result: ctx.prop?.nested
> 
> ```
> 
> 

I attempted this a while ago, but it didn't work because of the parentheses we add for location comments.

**Comment by amakhrov** [comment]
Oh I see! It would generate something like `(((ctx).a)?.b).c`, where the parens break the chain

**Comment by amakhrov** [comment]
It looks like the issue is much deeper than I imagined. Basically, the AST shape from the parser doesn't match the order of operations. Later the **compiler** reshapes this tree respecting long short-circuiting of `?.`.
(https://github.com/angular/angular/blob/a4ab6d6b72fe891f232d2446f8b7c454a82a5ba0/packages/compiler/src/compiler_util/expression_converter.ts#L564)

However, looks like the **typechecker** operates on the original (not transformed) AST. Hence, by the time Ast Visitor visits the SafePropertyRead node, it's too late to do anything - the chain has been already corrupted by those parentheses.

Perhaps this tree reshaping logic from the compiler should be duplicated in the typechecker. Or maybe perform it once - before even passing the AST to compiler and typechecker.

**Comment by JoostK** [comment]
> Perhaps this tree reshaping logic from the compiler should be duplicated in the typechecker. Or maybe perform it once - before even passing the AST to compiler and typechecker.

Indeed, the AST restructuring logic happens in the compiler's output emitter, i.e. the transform from "expression AST" to "output AST". The typechecker generates TypeScript AST nodes directly from "expression AST" nodes, so it hasn't gone through that transformation.

FYI I did attempt expanding the translation in the type-checker at one point but never got to finish it; I should have a look at that again.

**Comment by kbrilla** [comment]
any updates on this? 

I think I have similar issue

stackblitz: [link](https://stackblitz.com/edit/angular-ivy-gpknqn?file=src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fapp.component.html,tsconfig.json,src%2Fapp%2Fstring-only-pipe.pipe.ts,src%2Fmain.ts,src%2Fapp%2Fhello.component.ts)
```
export type AccessibleValue<T> =
  | { isAccessible: true; value: T }
  | {
      isAccessible: false;
      value: null;
    };
```
evaluates to string correctly in typescript
```
 if (this.testValue.accessible?.isAccessible) {
      this.stringOnlyToTest = this.testValue.accessible.value; 
    }
```
but in tempalte it throws 
```
Argument of type 'string | null' is not assignable to parameter of type 'string'.
Type 'null' is not assignable to type 'string'.
```
```
<p *ngIf="testValue.accessible?.isAccessible">
{{ testValue.accessible.value | stringOnlyPipe }} <span></span> 
</p>
```

**Comment by jrista** [comment]
Just wanted to chime in and state, I seem to be encountering this issue (and maybe something more extensive) with Angular 14. With `strictTemplate` set to `true` in my tsconfig, most typing in a heck of a lot of my templates breaks. If I set this setting to false, then the issue goes away. As I understand it reading https://angular.io/guide/template-typecheck, I am basically only getting partial type checking, down to two levels deep, rather than full type checking. We use optional chaining fairly extensively in our templates...

**Comment by mirobo** [comment]
Can someone confirm that my issue is the same issue?

https://stackblitz.com/edit/dn7xyz?file=src%2Fexample%2Ftable-row-binding-example.html,src%2Fexample%2Ftable-row-binding-example.ts,src%2Fexample%2Fmy-component.ts,src%2Fexample%2Fsome-data-for-table.ts

```
<ng-container *ngIf="dataField1.account?.id !== undefined">
  <my-component [dataField1]="dataField1.account.id"> </my-component>
</ng-container>
```


```
Error: src/example/table-row-binding-example.html:2:18 - error TS2322: Type 'number | undefined' is not assignable to type 'number'.
  Type 'undefined' is not assignable to type 'number'.

2   <my-component [dataField1]="dataField1.account.id"> </my-component>
                   ~~~~~~~~~~

  src/example/table-row-binding-example.ts:13:16
    13   templateUrl: 'table-row-binding-example.html',
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component TableRowBindingExample.


Error: src/example/table-row-binding-example.html:2:50 - error TS2532: Object is possibly 'undefined'.

2   <my-component [dataField1]="dataField1.account.id"> </my-component>
                                                   ~~

  src/example/table-row-binding-example.ts:13:16
    13   templateUrl: 'table-row-binding-example.html',
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component TableRowBindingExample.

```

Latest Angular version btw 😁.

Not sure this is even resolvable in the template. Of course I could define a fallback value with ` dataField1.account?.id ?? 0` but in some cases I don't want to display the component, if the data is missing and I cannot narrow the type down with an `*ngIf` it seems...

**Comment by PowerKiKi** [comment]
[This brand new StackBlitz](https://stackblitz.com/edit/stackblitz-starters-is4e1j?file=src%2Fmain.ts) demonstrate that the issue still persist with Angular 18.0 and the new control flow.

The code is:

```ts
type MyType = {
  foo: {
    bar: boolean;
  };
};

@Component({
  selector: 'app-root',
  standalone: true,
  template: `
    @if(value?.foo){
      this condition compile, it's OK
    }

    @if(value?.foo.bar){
      this condition should compile, but it does not
    }

    <@if(value?.foo.bar ?? true){
      this condition should compile even more, but it still does not
    }
  `,
})
export class App {
  public value: MyType | null = null;
}
````

And the one warning and two errors are incorrect and unexpected:

```
▲ [WARNING] NG8102: The left side of this nullish coalescing operation does not include 'null' or 'undefined' in its type, therefore the '??' operator can be safely removed. [plugin angular-compiler]

    src/main.ts:23:9:
      23 │     <@if(value?.foo.bar ?? true){
         ╵          ~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main.ts:19:19:
      19 │     @if(value?.foo.bar){
         ╵                    ~~~


✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main.ts:23:20:
      23 │     <@if(value?.foo.bar ?? true){
         ╵                     ~~~
```

**Comment by kbrilla** [comment]
I'm afraid that my hopes were rushed and my example from Angular v14 still don't work after updating to latest Angular and new code flow: 

 https://stackblitz.com/~/github.com/kbrilla/angular-issue-null-operator-v18.1

```ts
import { Component, Input, Pipe, PipeTransform } from '@angular/core';
import { bootstrapApplication } from '@angular/platform-browser';
import 'zone.js';

export type AccessibleValue<T> =
  | { isAccessible: true; value: T }
  | {
      isAccessible: false;
      value: null;
    };

@Pipe({
  standalone: true,
  name: 'stringOnlyPipe',
})
export class StringOnlyPipePipe implements PipeTransform {
  transform(value: string): string {
    return value;
  }
}

@Component({
  standalone: true,
  selector: 'my-app',
  imports: [StringOnlyPipePipe],
  template: `
    @if(testValue.accessible?.isAccessible === true) {
      <p>
      {{ testValue.accessible.value | stringOnlyPipe }} <span></span>
      </p>
    } 
    <p>
    {{ testValue.accessible?.value | stringOnlyPipe }} <span></span>
    </p>
`,
})
export class AppComponent {
  testValue: { accessible?: AccessibleValue<string> } = {
    accessible: { isAccessible: true, value: 'type narrowing test' },
  };
  stringOnlyToTest!: string;

  test() {
    // this.stringOnlyToTest = this.testValue.accessible.value; throws as it should

    if (this.testValue.accessible?.isAccessible) {
      this.stringOnlyToTest = this.testValue.accessible.value;
    }
  }
}

bootstrapApplication(AppComponent);

```



```

Application bundle generation failed. [0.079 seconds]

✘ [ERROR] NG2: Object is possibly 'undefined'. [plugin angular-compiler]

    src/main.ts:29:30:
      29 │       {{ testValue.accessible.value | stringOnlyPipe }} <span></span>
         ╵                               ~~~~~


✘ [ERROR] NG5: Argument of type 'string | null' is not assignable to parameter of type 'string'.
  Type 'null' is not assignable to type 'string'. [plugin angular-compiler]

    src/main.ts:29:30:
      29 │       {{ testValue.accessible.value | stringOnlyPipe }} <span></span>
         ╵                               ~~~~~


✘ [ERROR] NG5: Argument of type 'string | null | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'. [plugin angular-compiler]

    src/main.ts:33:7:
      33 │     {{ testValue.accessible?.value | stringOnlyPipe }} <span></span>
         ╵        ~~~~~~~~~~~~~~~~~~~~~~~~~~~


```

<img width="810" alt="image" src="https://github.com/angular/angular/assets/22288449/3d90af46-740b-4469-9f71-0ce138f01e4f">

**Comment by csvn** [comment]
Don't know if it deserves it's own issue either, but I find it confusing that `foo?.bar` is `null` in Angular templates, but `undefined` in JS when `foo` is `undefined` 🤔

**Comment by JoostK** [comment]
> Don't know if it deserves it's own issue either, but I find it confusing that `foo?.bar` is `null` in Angular templates, but `undefined` in JS when `foo` is `undefined` 🤔

Angular predates optional chaining in JS and choose different semantics. See #34385.

**Comment by kbrilla** [comment]
> > Don't know if it deserves it's own issue either, but I find it confusing that `foo?.bar` is `null` in Angular templates, but `undefined` in JS when `foo` is `undefined` 🤔
> 
> Angular predates optional chaining in JS and choose different semantics. See #34385.

Is there any change that in future version Angular team will try to align with TS on that? There could be migration used for older code that would allow to transform older code `x?.c` to `x != null && x.c` right?

If not maybe signal components would be a good place to align with typescript on `?.` so no older code would be broken in non signal components?

**Comment by csvn** [comment]
>Angular predates optional chaining in JS and choose different semantics. See https://github.com/angular/angular/issues/34385.
Yes, I know. I'm just so used to optional chaining now that I found the behavior confusing. 🤞for https://github.com/angular/angular/issues/34385 to pick up steam, would be great with new control flow updates to align closer with Javascript standards.

**Comment by mirobo** [comment]
I found a workaround, so I can at least use strictTemplates:

@if(nested?.optionalProp?.optionalProp; **as optionalProp**) { ...

```
@if(nested?.optionalProp?.optionalProp; as optionalProp) {
<p *ngIf="optionalProp.someOtherOptionalProperty">
  {{optionalProp.someOtherOptionalProperty}}
</p>
}```

**Comment by JeanMeche** [comment]
@mirobo you'd be better off with `@let optionalProp = nested?.optionalProp?.optionalProp`  in 18.1+ while we have better alternatives.

**Comment by mirobo** [comment]
> @mirobo you'd be better off with `@let optionalProp = nested?.optionalProp?.optionalProp` in 18.1+ while we have better alternatives.

Yes, absolutely, thanks for the advice! But I didn't migrate yet on 18.1.. still on 18.0.x :D. Other than that I think wrapping it with an if-clause can help with overview in a template if I don't want to split it into smaller components.

Best option would still be if Angular handles this for us with its magic 😁

**Comment by PowerKiKi** [comment]
With Angular 21.2.7, with the following code, in this [brand new StackBlitz](https://stackblitz.com/edit/stackblitz-starters-bghesuj9?description=An%20angular-cli%20project%20based%20on%20@angular/animations,%20@angular/common,%20@angular/compiler,%20@angular/core,%20@angular/forms,%20@angular/platform-browser,%20@angular/platform-browser-dynamic,%20@angular/router,%20core-js,%20rxjs,%20tslib%20and%20zone.js&file=src%2Fmain.ts&template=node&title=Angular%20Starter), the incorrect behavior still happens:

```ts
import { Component } from '@angular/core';
import { bootstrapApplication } from '@angular/platform-browser';

type MyType = {
  foo: {
    bar: boolean;
  };
};

@Component({
  selector: 'app-root',
  standalone: true,
  template: `
    @if(value?.foo){
      this condition compile, it's OK
    }

    @if(value?.foo.bar){
      this condition should compile, but it does not
    }

    <@if(value?.foo.bar ?? true){
      this condition should compile even more, but it still does not
    }
  `,
})
class App {
  public value: MyType | null = null;
}

bootstrapApplication(App);
```
